Decomps and meta registrations for upsample_nearest 1D / 2D / 3D

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91318
* #91317
* __->__ #91260

Adds decompositions and meta registrations for the 1D, 2D, and 3D implementations of `upsample_nearest`. All related OpInfo-based tests for AOTAutograd now pass.

jbschlosser committed:
Decomps and meta registrations for upsample_nearest 1D / 2D / 3D

[ghstack-poisoned]

jbschlosser commented on torch/_decomp/decompositions.py:
```diff
@@ -1983,6 +1983,131 @@ def get_scale_value(scales, idx):
+    # convert output to correct memory format, if necessary
+    memory_format = utils.suggest_memory_format(input)
+
+    # following "heuristic: only use channels_last path when it's faster than the contiguous path"
```
I stole this heuristic from the `upsample_bilinear2d` decomp below, but I'm not sure where it came from or whether it's correct here.

vfdev-5 replied:
@jbschlosser can you detail a bit more this heuristic ?

I could find a reference in cuda code, but I could not find there where channels first memory format is set explicitly:

https://github.com/pytorch/pytorch/blob/0acfe2ce096b874303f1325c5188c457d18d828d/aten/src/ATen/native/cuda/UpSampleNearest2d.cu#L270-L331

Here is a repro code that can show memory format inconsistency vs aten:
```python
import torch

device = "cuda"


x = torch.ones(2, 3, 32, 32, dtype=torch.float32, device=device).contiguous(memory_format=torch.channels_last)
print("Input:", x.shape, x.stride(), x.device, x.dtype)


expected_output = torch._C._nn.upsample_nearest2d(x, [12, 12])
print(
    "Output, Aten:",
    expected_output.shape,
    expected_output.is_contiguous(),
    expected_output.is_contiguous(memory_format=torch.channels_last),
)

fn = torch.compile(torch.ops.aten.upsample_nearest2d.default)
decomp_output = fn(x, [12, 12])
print(
    "Output, Decomposition:",
    decomp_output.shape,
    decomp_output.is_contiguous(),
    decomp_output.is_contiguous(memory_format=torch.channels_last),
)
```

```
Input: torch.Size([2, 3, 32, 32]) (3072, 1, 96, 3) cuda:0 torch.float32
Output, Aten: torch.Size([2, 3, 12, 12]) False True
Output, Decomposition: torch.Size([2, 3, 12, 12]) True False
```

jbschlosser committed:
Update on "Decomps and meta registrations for upsample_nearest 1D / 2D / 3D"


Adds decompositions and meta registrations for the 1D, 2D, and 3D implementations of `upsample_nearest`. All related OpInfo-based tests for AOTAutograd now pass.

[ghstack-poisoned]

jbschlosser committed:
Update on "Decomps and meta registrations for upsample_nearest 1D / 2D / 3D"


Adds decompositions and meta registrations for the 1D, 2D, and 3D implementations of `upsample_nearest`. All related OpInfo-based tests for AOTAutograd now pass.

[ghstack-poisoned]

ezyang commented on torch/_meta_registrations.py:
```diff
@@ -2012,6 +2012,76 @@ def meta_scatter_reduce__two(self, dim, index, src, reduce, include_self=True):
+    r = 1
+    for v in vs:
+        r *= v
+    return r
```
I swear we already have a copy of this function lol

jbschlosser replied:
True, I found one in `torch/jit/_shape_functions.py`

ezyang commented on torch/_decomp/decompositions.py:
```diff
@@ -1983,6 +1983,135 @@ def get_scale_value(scales, idx):
+    scales_w: Optional[float] = None,
+) -> Tensor:
+    h_indices, w_indices = _compute_upsample_nearest_indices(input, output_size)
+    result = input[:, :, h_indices, w_indices]
```
Huh, I did not realize our true implementations computed indices separately from dereferencing. Inductor should help a lot here!

ezyang reviewed (approved):
Nice work!

eellison commented on torch/_meta_registrations.py:
```diff
@@ -2012,6 +2012,76 @@ def meta_scatter_reduce__two(self, dim, index, src, reduce, include_self=True):
+        input.numel() != 0 or multiply_integers(input.size()[1:]),
+        lambda: "Non-empty 4D data tensor expected but got a tensor with sizes {input.size()}",
+    )
+    full_output_size = upsample_common_check(
```
.You need the same `following "heuristic: only use channels_last path when` logic here, and it should only be for cuda. Can you update this and the decomp to only do this on cuda ? also for the other decomp as well would be great

jbschlosser replied:
Sure, adding similar logic here

> also for the other decomp as well would be great

Sorry, which other decomp are you referring to?

jbschlosser replied:
@eellison do you know if there's a reason why we don't verify output `memory_format` in the decomp crossref testing?

eellison commented on torch/_decomp/decompositions.py:
```diff
@@ -1983,6 +1983,135 @@ def get_scale_value(scales, idx):
+    # convert output to correct memory format, if necessary
+    memory_format = utils.suggest_memory_format(input)
+
+    # following "heuristic: only use channels_last path when it's faster than the contiguous path"
```
this is only for the 2d implementation. [here](https://github.com/pytorch/pytorch/blob/07c61685c81820fa9c9d74cf879700d896bc49c3/aten/src/ATen/native/cuda/UpSampleNearest2d.cu#L239) and [here](https://github.com/pytorch/pytorch/blob/07c61685c81820fa9c9d74cf879700d896bc49c3/aten/src/ATen/native/cuda/UpSampleBilinear2d.cu#L288)

jbschlosser replied:
Cool, wasn't sure if there was an analogous channels_last3d thing going on. Will remove

eellison reviewed (commented):
nice !

jbschlosser committed:
Update on "Decomps and meta registrations for upsample_nearest 1D / 2D / 3D"


Adds decompositions and meta registrations for the 1D, 2D, and 3D implementations of `upsample_nearest`. All related OpInfo-based tests for AOTAutograd now pass.

[ghstack-poisoned]

jbschlosser committed:
Update on "Decomps and meta registrations for upsample_nearest 1D / 2D / 3D"


Adds decompositions and meta registrations for the 1D, 2D, and 3D implementations of `upsample_nearest`. All related OpInfo-based tests for AOTAutograd now pass.

[ghstack-poisoned]

jbschlosser commented:
Going to merge before the freeze; can address any other changes in a follow-up

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot rebase

jbschlosser committed:
Update on "Decomps and meta registrations for upsample_nearest 1D / 2D / 3D"


Adds decompositions and meta registrations for the 1D, 2D, and 3D implementations of `upsample_nearest`. All related OpInfo-based tests for AOTAutograd now pass.

[ghstack-poisoned]

jbschlosser committed:
Update on "Decomps and meta registrations for upsample_nearest 1D / 2D / 3D"


Adds decompositions and meta registrations for the 1D, 2D, and 3D implementations of `upsample_nearest`. All related OpInfo-based tests for AOTAutograd now pass.

[ghstack-poisoned]
